src/inmanta/execute/dataflow/root_cause.py

```diff
@@ -0,0 +1,134 @@
+
+
+class UnsetRootCauseAnalyzer:
+    """
+        Analyzes the root causes for attributes being unset among a collection of attribute nodes.
+        The main entrypoint for this class is the root_causes method.
+    """
+
+    def __init__(self, nodes: Iterable[AttributeNode]) -> None:
+        self.nodes: FrozenSet[AttributeNode] = frozenset(nodes)
+
+    def root_causes(self) -> Set[AttributeNode]:
+        """
+            Returns the root causes from this instances' set of attribute nodes. An attribute node c
+            is defined as the cause for an other attribute node n iff c being unset leads to n
+            being unset.
+            Formally, the relation is_cause(c, x) is defined by three rules:
+                1. is_cause(c, x) <- c in `x = c` in graph
+                    (If `x = c` then c is responsible for x receiving a value)
+                2. is_cause(c, x) <- exists y: is_cause(c, y) and is_cause(y, x)
+                    (Cause is transitive)
+                3. is_cause(c, x) <- exists i : is_index_attr(x, i) and is_cause(c, x.i)
+                    (If an index attribute of x is unset this blocks execution. If c is the cause for the index
+                        value being unset, it is the cause for x being unset)
+                4. is_cause(c, x) <- exists y, z : `x = y.z` in graph and is_cause(c, y)
+                    (If x refers to y.z but y is unset, this blocks execution. If c is the cause for y being unset,
+                        it is the cause for x being unset)
+
+                example (entity definitions omitted for clarity):
+                    model:
+                        index V(i)
+
+                        c = C()
+                        u = U()
+                        x = X()
+
+                        u.v = V(n = 42, i = c.i)
+                        x.n = u.v.n
+
+                    root_cause_analysis (capital letters refer to the single instance of that entity, not the entity itself):
+                        is_cause(c.i, x.n)
+                            <-(4)- `x.n = u.v.n` in graph and is_cause(c.i, u.v)
+                            <----- is_cause(c.i, u.v)
+                            <-(2)- is_cause(V, u.v) and is_cause(c.i, V)
+                            <-(1)- `u.v = V` in graph and is_cause(c.i, V)
+                            <----- is_cause(c.i, V)
+                            <-(3)- is_index_attr(V, i) and is_cause(c.i, V.i)
+                            <----- is_cause(c.i, V.i)
+                            <-(1)- `V.i = c.i` in graph
+                            <----- true
+            Rules 2 to 4 are implemented as propagation steps by
+            _assignment_step, _child_attribute_step and _parent_instance_step respectively.
+        """
+        causes: Set[AttributeNode] = set(())
+        for node in self.nodes:
+            others: FrozenSet[AttributeNode] = self.nodes.difference({node})
+            seen: Set[AssignableNode] = set(())
```

Comment:
The code still searches for relative cause (i.e. determine if one nodes is the caused  by any of the others)
It would be easier to search for roots (i.e. find the nodes that have no underlying cause). 

The relative searching requires the double loop and discards the cache (`seen`) on every iteration of outer loop. 

By searching for roots, it becomes much easier. 

(i.e. depth first search, building a cache of all roots and all non-roots, each node is handled exactly once, time complexity O(Nodes+Edges) )

Reply:
The difficulty at this point is that the propagation rules don't stop at AttributeNodes so searching for roots might produce AssignableNodes that are not AttributeNodes, which is at this point undesirable: the goal is to find the root cause among the given set. To do this we would need to keep track of paths taken which would make the algorithm more complex (read complexity, not time complexity).
If we really want to get rid of the double loop I think the best approach would be to come back on our earlier decision and resolve #1962 in this pull request as well. Unless I'm missing something.

Reply:
The double loop doesn't really bother me, it is the dropping of the cache (seen). 

If you don't mind, I'll take a shot at making a solution as well.

Reply:
Yes, feel free. I'm curious to see what you come up with.

Reply:
Add PR to this branch